src/main/kotlin/org/rust/ide/utils/ExprUtils.kt

```diff
@@ -0,0 +1,212 @@
+package org.rust.ide.utils
+
+import com.intellij.openapi.project.Project
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.RsElementTypes.ANDAND
+import org.rust.lang.core.psi.RsElementTypes.OROR
+
+/**
+ * Returns `true` if all elements are `true`, `false` if there exists
+ * `false` element and `null` otherwise.
+ */
+private fun <T> List<T>.allMaybe(predicate: (T) -> Boolean?): Boolean? {
+    val values = map(predicate)
+    val nullsTrue = values.all {
+        it ?: true
+    }
+    val nullsFalse = values.all {
+        it ?: false
+    }
+    return if (nullsTrue == nullsFalse) nullsTrue else null
+}
+
+/**
+ * Check if an expression is functionally pure
+ * (has no side-effects and throws no errors).
+ *
+ * @return `true` if the expression is pure, `false` if
+ *        > it is not pure (has side-effects / throws errors)
+ *        > or `null` if it is unknown.
+ */
+fun RsExpr.isPure(): Boolean? {
+    return when (this) {
+        is RsArrayExpr -> when (semicolon) {
+            null -> exprList.allMaybe(RsExpr::isPure)
+            else -> exprList[0].isPure() // Array literal of form [expr; size],
+        // size is a compile-time constant, so it is always pure
+        }
+        is RsStructExpr -> when (structExprBody.dotdot) {
+            null -> structExprBody.structExprFieldList
+                .map { it.expr }
+                .allMaybe { it?.isPure() } // TODO: Why `it` can be null?
+            else -> null // TODO: handle update case (`Point{ y: 0, z: 10, .. base}`)
+        }
+        is RsBinaryExpr -> when (operatorType) {
+            ANDAND, OROR -> listOf(left, right!!).allMaybe(RsExpr::isPure)
+            else -> null // Have to search if operation is overloaded
+        }
+        is RsTupleExpr -> exprList.allMaybe(RsExpr::isPure)
+        is RsFieldExpr -> expr.isPure()
+        is RsParenExpr -> expr.isPure()
+        is RsBreakExpr -> false // Changes execution flow
```

Comment:
You can write condense this a little, like this: `is RsTupleExpr, is RsFieldExpr` ...